### PR TITLE
[JAVA_API][GSOC] Add Java API bindings

### PR DESCRIPTION
**Pull request by rajatkrishna**
## Details

This PR adds JNI wrappers to the OpenVINO Java API and includes a `hello_query_device` sample that demonstrates the usage of the Query Device API feature to print device configuration and default config values.

Below is a list of JNI wrappers added:

* Read and compile a model directly from the model path

    | ov::Core | Core |
    |----------|------|
    | compile_model(model_path) | compile_model(modelPath) |
    | compile_model(model_path, device_name) | compile_model(modelPath, deviceName) |
    | compile_model(model_path, device_name, properties={}) | compile_model(modelPath, deviceName, properties) |

* Compile an ov.Model with the given properties
    | ov::Core | Core |
    |----------|------|
    | compile_model(model, device_name, properties={}) | compile_model(model, deviceName, properties) |
   
* List devices available for inference

    | ov::Core | Core |
    |----------|------|
    | get_available_devices() | get_available_devices() |

* Extract data from ov.Any type
    | ov::Any | Any |
    |----------|------|
    |  | asString() |
    |  | asList() |

* List the input and output tensor information of a CompiledModel. Returns the tensor shape, name and element type of each input/output tensor. 

    | ov::CompiledModel | CompiledModel|
    |----------|------|
    | inputs() | inputs() |
    | outputs() | outputs() |

* Set input/output tensors to infer on by tensor name when the model has multiple input/output tensors

    | ov::InferRequest| InferRequest |
    |----------|------|
    | set_tensor(tensor_name, tensor) | set_tensor(tensorName, tensor) |

* Create integer tensors and extract data as an int array

    | ov::Tensor| Tensor|
    |----------|------|
    | Tensor(shape, data) | Tensor(shape, data) |
    | data() | as_int() |

* Create Long tensors 

    | ov::Tensor| Tensor|
    |----------|------|
    | Tensor(shape, data) | Tensor(shape, data) |

* Save model into IR format from memory

    | | |
    |----------|------|
    | ov::serialize(model, xml_path, bin_path) | Openvino.serialize(model, xmlPath, binPath) |

### CODE REVIEW

File: modules/java_api/src/main/cpp/jni_common.hpp [hunk 1]
```diff
@@ -202,3 +202,33 @@ static std::vector<size_t> jintArrayToVector(JNIEnv *env, jintArray dims)
 
     return std::vector<size_t>();
 }
+
+template<typename T>
```
Comment (likholat) [inline modules/java_api/src/main/cpp/jni_common.hpp:206]:
```suggestion
```

File: modules/java_api/src/main/cpp/jni_common.hpp [hunk 2]
```diff
@@ -202,3 +202,33 @@ static std::vector<size_t> jintArrayToVector(JNIEnv *env, jintArray dims)
...
+static jobject vectorToJavaList(JNIEnv *env, std::vector<T> items)
```
Comment (likholat) [inline modules/java_api/src/main/cpp/jni_common.hpp:207]:
```suggestion
static jobject vectorToJavaList(JNIEnv *env, std::vector<std::string> items)
```

File: modules/java_api/src/main/cpp/openvino.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#include "jni_common.hpp"
+
+using namespace ov;
+
+JNIEXPORT void JNICALL Java_org_intel_openvino_Openvino_serialize(JNIEnv *env, jobject obj, jlong modelAddr, jstring xmlPath, jstring binPath)
+{
+    JNI_METHOD("serialize",
+        std::string xml_path = jstringToString(env, xmlPath);
+        std::string bin_path = jstringToString(env, binPath);
+        std::shared_ptr<const Model> *model = reinterpret_cast<std::shared_ptr<const Model> *>(modelAddr);
```
Comment (likholat) [inline modules/java_api/src/main/cpp/openvino.cpp:18]:
```suggestion
        std::shared_ptr<const Model> model = std::make_shared<const Model>(modelAddr);
```

File: modules/java_api/src/main/cpp/openvino.cpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+        serialize(*model, xml_path, bin_path);
```
Comment (likholat) [inline modules/java_api/src/main/cpp/openvino.cpp:20]:
```suggestion
        serialize(model, xml_path, bin_path);
```

File: modules/java_api/src/main/cpp/jni_common.hpp [hunk 5]
```diff
@@ -202,3 +202,33 @@ static std::vector<size_t> jintArrayToVector(JNIEnv *env, jintArray dims)
...
+    static const char method_name[] = "vectorToJavaList";
+    try
+    {
+        jclass arrayClass = env->FindClass("java/util/ArrayList");
+        jmethodID arrayInit = env->GetMethodID(arrayClass, "<init>", "()V");
+        jobject arrayObj = env->NewObject(arrayClass, arrayInit);
+        jmethodID arrayAdd = env->GetMethodID(arrayClass, "add", "(Ljava/lang/Object;)Z");
+
+        for (const auto& item : items) {
+            jstring string = env->NewStringUTF(item.c_str());
```
Comment (likholat) [inline modules/java_api/src/main/cpp/jni_common.hpp:218]: There would be an error if item is not `std::string`

File: modules/java_api/src/main/cpp/core.cpp [hunk 6]
```diff
@@ -76,6 +76,73 @@ JNIEXPORT jlong JNICALL Java_org_intel_openvino_Core_CompileModel(JNIEnv *env, j
...
+    JNI_METHOD("CompileModel3",
+        std::string n_device = jstringToString(env, device);
+        std::string n_path = jstringToString(env, path);
+        AnyMap map;
+        for (const auto& it : javaMapToMap(env, props)) {
+            map[it.first] = it.second;
+        }
+
+        Core *core = (Core *)coreAddr;
+        CompiledModel *compiled_model = new CompiledModel();
```
Comment (likholat) [inline modules/java_api/src/main/cpp/core.cpp:119]: Just for understanding in future :)
```suggestion
        // AnyMap will be copied inside compile_model, so we don't have to track this object lifetime
        CompiledModel *compiled_model = new CompiledModel();
```

File: modules/java_api/src/main/cpp/openvino.cpp (hunk 3 above)
Comment (likholat) [reply modules/java_api/src/main/cpp/openvino.cpp]: My mistake, not needed

File: modules/java_api/src/main/cpp/jni_common.hpp (hunk 5 above)
Comment (rajatkrishna) [reply modules/java_api/src/main/cpp/jni_common.hpp:217]: [Updated](https://github.com/openvinotoolkit/openvino_contrib/pull/709/commits/98c3e6debff015a71afd196771698d99b048278b#diff-7eea2d8fe6a6c9d40a378141adc603ea0b429492bd1410ae550edd7cd2d31807) to accept `std::string` items.

**Comment by redradist** [comment]
What is the purpose of this PR if `openvino_contrib` has `java_api` already

**Comment by likholat** [comment]
@rajatkrishna please run [google-java-format](https://github.com/google/google-java-format) locally before commit: 
```
cd openvino_contrib/modules/java_api
find . -name "*.java" -type f -print | xargs java -jar google-java-format-1.18.1-all-deps.jar -i -a
```

**Comment by likholat** [comment]
@ilya-lavrenov could you merge this PR?

**Pull request merged**